File: src/main/java/com/urbanhunt/app/model/Challenge.java
package com.urbanhunt.app.model;

import com.fasterxml.jackson.annotation.JsonFormat;
import com.google.cloud.firestore.annotation.DocumentId;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Challenge {

    @DocumentId
    private String id;

    private String title;

    private ChallengeStatus status;

    private String country;

    private String cityName;

    private String createdBy; // Email of user who created the challenge

    private String prizePhotoUrl; // Optional photo of the prize

    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd'T'HH:mm:ss")
    private Date createdAt;

    @Builder.Default
    private List<Hint> hints = new ArrayList<>();

    private Completion completion;

    private Long commentsCount;

    public enum ChallengeStatus {
        ACTIVE,
        COMPLETED,
        ARCHIVED
    }

}

File: app/ios/UrbanHunt/Models/Challenge.swift
//
//  Challenge.swift
//  UrbanHunt
//
//  Challenge model
//

import Foundation

struct Challenge: Identifiable, Codable {
    let id: String
    let title: String
    let status: ChallengeStatus
    let country: String
    let cityName: String
    let createdBy: String?
    let creator: UserSummary?
    let location: String? // Private note for creator only
    let prizePhotoUrl: String?
    let createdAt: Date
    let hints: [Hint]?
    let completion: Completion?
    var commentsCount: Int?
    let nextHintDate: Date?
    let confirmationId: String? // Prize confirmation ID for QR code

    enum ChallengeStatus: String, Codable {
        case draft = "DRAFT"
        case active = "ACTIVE"
        case completed = "COMPLETED"
        case archived = "ARCHIVED"
    }
}

struct UserSummary: Codable {
    let email: String
    let name: String
    let pictureUrl: String?
    let socialMediaUrl: String?
}

struct Hint: Codable {
    let content: String
    let link: String?
    let publishedAt: Date?
}

struct Completion: Codable {
    let userId: String
    let userName: String?
    let completedAt: Date
    let qrCodeScanned: String?
}